File: cpp/longest-increasing-subsequence.cpp
#include <iostream>
#include <cstdio>
#include <algorithm>
#include <vector>

using namespace std;

class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
		size_t N = nums.size();
		if (N < 1) return 0;
		vector<int> dp(N+1, 1);
		dp[0] = 0;

		int res = 1;
		for (int i = 1; i < N; ++i) {
			for (int j = 0; j < i; ++j) {
				if (nums[i] > nums[j]) dp[i+1] = max(dp[i+1], dp[j+1]+1);
			}
			res = max(res, dp[i+1]);
		}

		return res;
    }
};

int main(int argc, char *argv[]) {
	Solution s;
	{
		vector<int> nums = {10, 9, 2, 5, 3, 7, 101, 18};
		cout << s.lengthOfLIS(nums) << endl;
	}
	{
		vector<int> nums = {10,9,2,5,3,4};
		cout << s.lengthOfLIS(nums) << endl;
	}
	return 0;
}

File: js/longest-increasing-subsequence.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var lengthOfLIS = function(nums) {
	var N = nums.length;
	if (N <= 1) return N;
	var dp = new Array(N);

	dp[0] = 1;
	var res = 1;
	for (var i = 1; i < N; i++) {
		dp[i] = 1;
		for (var j = 0; j < i; j++) {
			if (nums[i] > nums[j]) dp[i] = Math.max(dp[i], dp[j]+1);
		}
		res = Math.max(res, dp[i]);
	}

	return res;
};

module.exports = lengthOfLIS;
